src/Datadog.Trace.DuckTyping/ILHelpers.cs

```diff
@@ -0,0 +1,296 @@
+        {
+            switch (value)
+            {
+                case 0:
+                    il.Emit(OpCodes.Ldc_I4_0);
+                    break;
+                case 1:
+                    il.Emit(OpCodes.Ldc_I4_1);
+                    break;
+                case 2:
+                    il.Emit(OpCodes.Ldc_I4_2);
+                    break;
+                case 3:
+                    il.Emit(OpCodes.Ldc_I4_3);
+                    break;
+                case 4:
+                    il.Emit(OpCodes.Ldc_I4_4);
+                    break;
+                case 5:
+                    il.Emit(OpCodes.Ldc_I4_5);
+                    break;
+                case 6:
+                    il.Emit(OpCodes.Ldc_I4_6);
+                    break;
+                case 7:
+                    il.Emit(OpCodes.Ldc_I4_7);
+                    break;
+                case 8:
+                    il.Emit(OpCodes.Ldc_I4_8);
+                    break;
+                default:
+                    il.Emit(OpCodes.Ldc_I4_S, value);
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Convert a current type to an expected type
+        /// </summary>
+        /// <param name="il">ILGenerator</param>
+        /// <param name="actualType">Actual type</param>
+        /// <param name="expectedType">Expected type</param>
+        internal static void TypeConversion(ILGenerator il, Type actualType, Type expectedType)
+        {
+            if (actualType == expectedType)
+            {
+                return;
+            }
+
+            if (actualType.IsGenericParameter && expectedType.IsGenericParameter)
+            {
+                return;
+            }
+
+            var actualUnderlyingType = actualType.IsEnum ? Enum.GetUnderlyingType(actualType) : actualType;
+            var expectedUnderlyingType = expectedType.IsEnum ? Enum.GetUnderlyingType(expectedType) : expectedType;
+
+            if (actualUnderlyingType.IsValueType)
+            {
+                if (expectedUnderlyingType.IsValueType && actualUnderlyingType != expectedUnderlyingType)
+                {
+                    // If both underlying types are value types then both must be of the same type.
+                    throw new InvalidCastException();
+                }
+                else if (!expectedUnderlyingType.IsValueType)
+                {
+                    // An underlying type can be boxed and converted to an object or interface type if the actual type support this
+                    // if not we should throw.
+                    if (expectedUnderlyingType == typeof(object) || expectedUnderlyingType.IsAssignableFrom(actualUnderlyingType))
+                    {
+                        il.Emit(OpCodes.Box, actualType);
+                        il.Emit(OpCodes.Castclass, expectedType);
+                    }
+                    else if (expectedUnderlyingType.IsAssignableFrom(actualUnderlyingType))
```

Comment:
This `expectedUnderlyingType.IsAssignableFrom(actualUnderlyingType)` exactly matches the condition in the previous if clause. Is this supposed to be the case and maybe the above if clause is supposed to be an `&&` condition?

Reply:
You're totally right, the branching logic of the if/else was wrong, I changed the implementation of that method in https://github.com/DataDog/dd-trace-dotnet/pull/946/commits/e3020964dcbb2b80b94368777abbdd5b1de4e248